### ISSUE TITLE
JIT session error triggered by precompilation

### ISSUE THREAD

**Issue by pxl-th** [JuliaGPU/GPUCompiler.jl#611]
Using PrecompileTools.jl to [precompile](https://github.com/JuliaNeuralGraphics/GaussianSplatting.jl/blob/main/ext/GaussianSplattingCUDAExt/GaussianSplattingCUDAExt.jl) GaussianSpaltting.jl triggers following error on 1.10 when executing code after precompilation:

```julia
JIT session error: Symbols not found: [ julia.gpu.state_getter ]
JIT session error: Failed to materialize symbols: { (JuliaOJIT, { julia_#throw_inexacterror_2603 }) }
JIT session error: Failed to materialize symbols: { (JuliaOJIT, { julia_#GaussianRasterizer#16_2598 }) }
```

MWE:

The error seems to be triggered at the moment of calling `GaussianRasterizer` ctor.

```julia
using CUDA
using cuDNN
using GaussianSplatting

function main()
    kab = GaussianSplatting.gpu_backend()

    points = CuArray(rand(Float32, 3, 128))
    colors = CuArray(rand(Float32, 3, 128))
    scales = CuArray(rand(Float32, 3, 128))

    camera = GaussianSplatting.Camera(; fx=100f0, fy=100f0, width=256, height=256)
    gaussians = GaussianSplatting.GaussianModel(points, colors, scales; max_sh_degree=0)
    rasterizer = GaussianSplatting.GaussianRasterizer(kab, camera; auxiliary=false)
    return
end
main()
```

**Comment by maleadt** [comment]
cc @vchuravy

**Comment by KSepetanc** [comment]
PackageCompiler fails to compile CUDA with `compile=all`. As suggested in [Discourse](https://discourse.julialang.org/t/packagecompiler-fails-to-create-app-for-madnlpgpu-examodels-cudss-linear-solver/125998), this issue might be related or the root of it.

**Comment by KSepetanc** [comment]
I wish to raise the importance of this issue as it blocks compilation of any CUDA.jl related project (using both PrecompileTools and PackageCompiler).

**Comment by maleadt** [comment]
MWE:

```julia
module MWE

using GPUCompiler, PrecompileTools

struct CompilerParams <: AbstractCompilerParams end

@inline threadfence_block() = ccall("llvm.nvvm.membar.cta", llvmcall, Cvoid, ())
kernel() = threadfence_block()

@setup_workload begin
    source = methodinstance(typeof(kernel), Tuple{})
    target = PTXCompilerTarget(cap=v"5.0")
    params = CompilerParams()
    config = CompilerConfig(target, params)
    job = CompilerJob(source, config)

    @compile_workload begin
        JuliaContext() do ctx
            GPUCompiler.compile(:llvm, job)
        end
    end
end

end # module MWE
```

```
Precompiling MWE...
LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.membar.cta
```

**Comment by KSepetanc** [comment]
It is either random or it got worse with Julia 1.12. I am now getting this error with default settings when compiling with PackageCompiler. Previously I got error only with `--compile=all`.

Should we open issue on Julia github as probably solution should be seeked there?

**Comment by KSepetanc** [comment]
I was unable to recreate failure running PrecompileTools and Julia 1.12 even when using full scale test case with CUDA (on 1.11 it fails).

However, PackageCompiler and JuliaC still fail with 1.12.
The first with `LLVM ERROR: Broken module found` error and function being call with register value rather than compile time constant. The second errors with `LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.barrier0` as stated above.

This led me to try again with different approach. We already know creating system image with `CUDA.jl` fails (Cannot select error) so I tried to narrow it down. Including the GPUCompiler in the system image it self already fails (`create_sysimage(["GPUCompiler"]; sysimage_path="ExampleSysimage.so")`). The error from below led me to check src/runtime.jl, in particular `function compile` (lines 69-90). There is FIXME sign that appears to be related to this. Including CUDA to system image fails as already including GPUCompiler fails. Could it be that this is the remaining culprit (at least on v1.12)?

```
    # FIXME: if the function is a symbol, implying it will be specified by the target,
    #        we won't be able to call this function here or we'll get UndefVarErrors.
    #        work around that by generating an llvmcall stub. can we do better by
    #        using the new nonrecursive codegen to handle function lookup ourselves?
```

```
C:/Users/karlo/.julia/artifacts/b17bda08a19173572926f43a48aad5ef3d845e7c/extracted_files/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\karlo\AppData\Local\Temp\jl_4VIdiJ41cl-o.a(text#0.o): in function `julia_signal_exception_106257':
C:\Users\karlo\.julia\packages\GPUCompiler\j4HFa\src/runtime.jl:85:(.text+0x7c3d6): undefined reference to `gpu_signal_exception'
C:/Users/karlo/.julia/artifacts/b17bda08a19173572926f43a48aad5ef3d845e7c/extracted_files/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\karlo\AppData\Local\Temp\jl_4VIdiJ41cl-o.a(text#7.o): in function `malloc':
C:\Users\karlo\.julia\packages\GPUCompiler\j4HFa\src/runtime.jl:85:(.text+0x147a6a): undefined reference to `gpu_malloc'
C:/Users/karlo/.julia/artifacts/b17bda08a19173572926f43a48aad5ef3d845e7c/extracted_files/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\karlo\AppData\Local\Temp\jl_4VIdiJ41cl-o.a(text#7.o): in function `report_oom':
C:\Users\karlo\.julia\packages\GPUCompiler\j4HFa\src/runtime.jl:85:(.text+0x147a7e): undefined reference to `gpu_report_oom'
collect2.exe: error: ld returned 1 exit status
```

**Comment by apozharski** [comment]
> I know it is not standard to use AI for these things, but I needed to since I am not an expert in this. I used Gemini to try to make MWE. What it did is create counter example without GPUCompiler that actually needs to crash, but it does not, meaning compiler silently skips instructions potentially causing much more serious issue than leaked GPU instructions, i.e. silent no operation. I was skeptical about this so I crosschecked by asking ChatGPT without code remarks to say what should happen with code and it said it should crash (but it does not even when using -O0 setting). Feedback is much appreciated on this.
> 
> ```
> module IntrinsicLeakMWE
> 
> using InteractiveUtils
> 
> # 1. Define the NVPTX-specific intrinsic
> @inline function threadfence_block()
>     Base.llvmcall(
>         """
>         ; This intrinsic should only exist on the NVPTX (GPU) backend.
>         call void @llvm.nvvm.membar.cta()
>         ret void
>         declare void @llvm.nvvm.membar.cta()
>         """,
>         Cvoid, Tuple{}
>     )
> end
> 
> function kernel()
>     threadfence_block()
>     return nothing
> end
> 
> # 2. Crash Trigger (Execute immediately upon include)
> 
> println("\n--- Attempting to force host compilation of GPU intrinsic ---")
> 
> try
>     # This line forces the full compiler pipeline, where the host LLVM backend fails
>     # when it encounters the GPU-only instruction.
>     InteractiveUtils.code_native(devnull, kernel, Tuple{})
>     
>     # If the code reaches here, the bug is the silent discard
>     println("🚨 WARNING: The compiler successfully ignored the intrinsic instead of crashing.")
>     
> catch e
>     if e isa Base.Exception && contains(sprint(showerror, e), "llvm.nvvm.membar.cta")
>         # This is the desired crash and the bug proof
>         println("\n✅ CRASH HIT: LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.membar.cta")
>     else
>         println("\n❌ UNEXPECTED ERROR during compilation.")
>         showerror(stdout, e)
>     end
>     rethrow(e)
> end
> 
> end # module IntrinsicLeakMWE
> ```

I will spare you the diatribe about how the practice of adding stochastic noise to the process of open source development is unhelpful and wastes peoples time, but forgive me my frustrated tone. 

As per usual, the stochastic parrots masquerading as intelligence you are using are being confidently incorrect. Had you checked and not redirected the `code_native` output to `devnull` you would have seen that the llvm code is not being compiled because it is trying to declare the intrinsic as a new function for some reason:
```assembly
--- Attempting to force host compilation of GPU intrinsic ---
        .text
        .file   "kernel"
        .section        .ltext,"axl",@progbits
        .globl  julia_kernel_2800               # -- Begin function julia_kernel_2800
        .p2align        4, 0x90
        .type   julia_kernel_2800,@function
julia_kernel_2800:                      # @julia_kernel_2800
; Function Signature: kernel()
; ┌ @ /home/anton/test/libmad/ihateai.jl:18 within `kernel`
# %bb.0:                                # %top
        push    rbp
; │ @ /home/anton/test/libmad/ihateai.jl:19 within `kernel`
; │┌ @ /home/anton/test/libmad/ihateai.jl:7 within `threadfence_block`
        movabs  rdi, offset ".L_j_str_Failed to parse LLVM asse...#1"
        movabs  rax, offset ijl_error
        mov     rbp, rsp
        call    rax
.Lfunc_end0:
        .size   julia_kernel_2800, .Lfunc_end0-julia_kernel_2800
; └└
                                        # -- End function
        .type   ".L_j_str_Failed to parse LLVM asse...#1",@object # @"_j_str_Failed to parse LLVM asse...#1"
        .section        .rodata.str1.1,"aMSl",@progbits,1
".L_j_str_Failed to parse LLVM asse...#1":
        .asciz  "Failed to parse LLVM assembly: \n<string>:5:1: error: expected instruction opcode\ndeclare void @llvm.nvvm.membar.cta()\n^\n"
        .size   ".L_j_str_Failed to parse LLVM asse...#1", 121

        .section        ".note.GNU-stack","",@progbits
🚨 WARNING: The compiler successfully ignored the intrinsic instead of crashing.
Main.IntrinsicLeakMWE
```
This is generally illegal to do in a define scope, which the string passed to `Base.llvmcall` is wrapped in (see [the julia documentation](https://docs.julialang.org/en/v1/base/c/#Core.Intrinsics.llvmcall), and [the llvm documentation](https://llvm.org/docs/LangRef.html#functions)).

If you remove `declare void @llvm.nvvm.membar.cta()` then you get hard crash as expected:
```
--- Attempting to force host compilation of GPU intrinsic ---
LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.membar.cta

[263999] signal 6 (-6): Aborted
```

In other words, nothing is being skipped, the bug we are trying to find is still at large, and plausible string generators reveal themselves to be nothing but that, unthinking plausible string generators.

**Comment by maleadt** [comment]
Please take a stab if you have the time; I'm currently preoccupied with non-GPU work.

**Comment by KSepetanc** [comment]
@apozharski this issue resurfaces with Julia 1.13. Could you try it with your use case?

**Comment by apozharski** [comment]
Yea it is very possible it does.

I can try some time to just confirm that the issue is back, though, to be frank, I don' t have the time or energy to keep up with the torrent of AI generated code in the julia ecosystem 😅, so I don't have much motivation to hunt down these esoteric bugs if libMad still works on 1.12 and its current set of dependencies.

**Comment by maleadt** [comment]
Curiously, this is caused by how `@doc` expands differently on 1.13. I'll work on some fixes.

**Comment by apozharski** [comment]
Ah fun you beat me to it 😅 . I found the unmarked functions last night but the documentation issue confused me because I for some reason didn't believe my lying eyes telling me that only the documented functions were in the ir. Thanks for looking at it @maleadt !

**Comment by apozharski** [comment]
Would it be good to set up a test in `CUDA.jl` using `JuliaC` to compile a basic MWE? Might help minimize these regressions. I can think a bit on what the best way to do this would be if you'd like @maleadt (or you can set something up if you have time!)

**Comment by maleadt** [comment]
Yeah I'll look into adding a test.

EDIT: added in https://github.com/JuliaGPU/GPUCompiler.jl/pull/911 and https://github.com/JuliaGPU/CUDA.jl/pull/3249

**Comment by maleadt** [comment]
This is covered by tests now, so I think we can close it.
